### PR TITLE
Initialize full Symfony layout in Back Office

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add new Symfony layout
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | CI should be green and so should:<br>Regular UI tests (no symfony layout option available) https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5327467076<br>UI tests with symfony layout https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5332273359 (still 10 campaigns failing, to be fixed later)<br>UI tests with explicit legacy layout (no regression) https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5328644737
| Fixed ticket?     | ~
| Related PRs       | ~
| Sponsor company   | ~

### How to test
To test the full UI tests campaign a new option was added to our tool in this PR https://github.com/PrestaShop/ga.tests.ui.pr/pull/30 this adds a new input in the launch modal that allows selecting which layout the back office will run on by switching the related feature flag via a CLI command
This should be used in all future PRs related to Symfony layout

Capture d’écran 2023-05-02 à 11 15 35

See example: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/4859492012/jobs/8662193087

### POC Result

This PR is only a POC a lot of things still need to be improved/done an Epic has been created https://github.com/PrestaShop/PrestaShop/issues/32875 to summarize all the remaining work to do.

### CODE REVIEW

**Comment by jolelievre** [comment]
Running whole campaign with prod mode + symfony layout: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5002594655
Running whole campaign with dev mode + symfony layout: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5002602121

File: src/PrestaShopBundle/Resources/views/Admin/Layout/default_layout.html.twig [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+      <nav id="header_infos" class="main-header">
+        <button class="btn btn-primary-reverse onclick btn-lg unbind ajax-spinner"></button>
+
+        {# Logo #}
+        <i class="material-icons js-mobile-menu">menu</i>
+        <a id="header_logo" class="logo float-left" href="{{ default_tab_link }}"></a>
+        <span id="shop_version">{{ ps_version }}</span>
+
+        <div class="component" id="quick-access-container">
+          {{ render(controller('PrestaShopBundle\\Controller\\Admin\\LegacyController::smartyComponentAction', {
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/views/Admin/Layout/default_layout.html.twig]: I'd use a twig function instead of rendering a controller for performance reasons in all components.

{{ smarty(template, variables) }}

File: src/Adapter/Tab/Repository/TabRepository.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Tab\Repository;
+
+use Doctrine\DBAL\Connection;
+use PrestaShop\PrestaShop\Core\Domain\Tab\ValueObject\TabId;
+
+class TabRepository
```
Comment (FabienPapet) [inline src/Adapter/Tab/Repository/TabRepository.php:34]: This should be placed inside prestashop bundle

File: src/PrestaShopBundle/Bridge/AdminController/ControllerConfiguration.php [hunk 3]
```diff
@@ -389,4 +413,16 @@ public function getUser(): Employee
     {
         return $this->user;
     }
+
+    /**
+     * Prepare configuration values that can be deduced from initial constructor parameters.
+     */
+    private function build(): void
+    {
+        if (strpos('Admin', $this->legacyControllerName) === 0) {
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Bridge/AdminController/ControllerConfiguration.php]: str starts with

File: src/PrestaShopBundle/Resources/config/services/adapter/tab.yml [hunk 4]
```diff
@@ -8,3 +8,9 @@ services:
       - '@doctrine.orm.entity_manager'
     tags:
       - { name: 'tactician.handler', command: 'PrestaShop\PrestaShop\Core\Domain\Tab\Command\UpdateTabStatusByClassNameCommand' }
+
+  PrestaShop\PrestaShop\Adapter\Tab\Repository\TabRepository:
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/adapter/tab.yml:12]: shoud be moved as the service should be placed inside Prestashop bundle

File: src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            'use_symfony_layout' => $useSymfonyLayout,
+        ];
+    }
+
+    private function renderSmartyContent(ControllerConfiguration $controllerConfiguration): array
+    {
+        $smarty = $this->context->getSmarty();
+        $smarty->assign($controllerConfiguration->templateVars);
+        $smarty->setTemplateDir([
+            _PS_BO_ALL_THEMES_DIR_ . 'new-theme' . DIRECTORY_SEPARATOR . 'template',
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php:157]: No constants inside Prestashop bundle

File: src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            $controllerConfiguration = $request->attributes->get(ContextShopListener::CONTROLLER_CONFIGURATION_ATTRIBUTE);
+            if ($controllerConfiguration instanceof ControllerConfiguration) {
+                foreach ($this->configurators as $configurator) {
+                    $configurator->configure($controllerConfiguration);
+                }
+                $layoutVariables = $controllerConfiguration->templateVars + $this->renderSmartyContent($controllerConfiguration);
+            }
+        }
+
+        return $layoutVariables + [
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php:103]: This is super dangerous, that means everyone can use on of this variable is the feature flag is set to on.


There's an alternative. You can create a twig extension that provides a method `getLegacyVars` that will require controller configurations.
in the beginning of the legacy template: `{% set ps_vars = getLegacyVars() %}` and then replace every var by `ps_vars.XXX``

WDYT ?

**Review by FabienPapet — CHANGES_REQUESTED** [review]

File: src/Adapter/Tab/Repository/TabRepository.php (hunk 2 above)
Comment (jolelievre) [reply src/Adapter/Tab/Repository/TabRepository.php:34]: The repository ADR and ne repositories are not ready yet, this will be refactored later

File: src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php (hunk 6 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php:103]: Hmm, the beginning of a solution I don't think prefixing `ps_vars` everywhere is gonna help us moving forward fast on this feature for the initial POC

I do agree that all these variables should be cleaned, the idea is to migrate each component piece by piece and clean the related variables so that only the component that need them access them Then we'll have to go through all the variables added via the Hydrator services check what's their usage and see if they are relevant to be in global variables (some are) then they will be properly moved into the extension Or if they are local variables needed for specific templates only (then we'll need to see how they can be injected properly in the template without being accessible everywhere)

But as for this POC I don't think you proposed alternative is acceptable 🤔

Comment (jolelievre) [reply src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php:103]: Also, keep in mind that this feature is still very much experimental, only activated via a feature flag, of course before the feature is fully released all this will have to be cleaned!! But we need a first step to start working with many people on parallel tasks However you can add a task explaining your worries and what should be done later in the Epic to make sure we don't forget about it https://github.com/PrestaShop/PrestaShop/issues/32875

File: src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php (hunk 5 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php:157]: @FabienPapet I tried to add this correctly but it has too many side effects, create-test-db install cli (an so all UI tests) fail because the const is not correctly defined in those context Mostly because it depends on ADMIN_DIR

So I think it should be handled in a dedicated PR

File: src/Adapter/Tab/Repository/TabRepository.php (hunk 2 above)
Comment (FabienPapet) [reply src/Adapter/Tab/Repository/TabRepository.php:34]: This is not about ADR, there's no legacy here, only doctrine

**Review by FabienPapet — CHANGES_REQUESTED** [review]
I do agree that the proposition I made is not ideal, and there should be a better one that we can try to find together. But I'm definitely against merging as it is for various reasons, but the main one is there's no temporary code inside prestashop. And these bad design already costs us a lot.

Perhaps we should start thinking about how we want it to be in the next versions, instead of trying to "adapt" the legacy. Maybe the legacy should start calling twig, instead of the opposite. After all, we're migrating admin pages to twig, why should twig use smarty ? The opposite would be more meanful and also avoids creating glue hell, as we always did.

Keep in mind that the main principle of a migration is to get rid of the legacy code, and by making everything compatible with smarty, we are doing exactly the opposite.Following the migration path we are using, there should be no more legacy controllers once they have been migrated, so we should not think about making legacy controllers compatible as they'll be deleted.

I can give an example of a migration path

- Design in symfony / twig what the future menu will look like.
- Design in symfony / twig what the future notification system will look like.
- Make a bridge to make this system compatible with smarty (and not the opposite)

That would reduce smarty usage to it's minimum until we can replace it.

At this point, making a bridge for smarty is just adding more coupling to smarty and the legacy.

**Comment by jolelievre** [comment]
@FabienPapet I think you didn't listen to what I explained during our meeting or I didn't explain properly so here's a summary:
- We can't handle the full symfony mayout in a single PR because it represents too much work
- We can't work on a single PR because it means only one person is working on the topic hen it could be split and handled by multiple persons
- This PR is only a POC or MVP for the final symfony rendering
- Initially, I didn't even intend to handle the smarty components but only to display a Symfony page with its content and nothing more, then each piece would be added PR by PR
- The problem with this is that we can't use the UI tests which are a huge help to detect regressions
- So to make a functional first POC the idea was to still display the other components and the easier way to render them the same way is to use the existing smarty templates that are also compatible with the UI tests
- The result is that even with the symfony layout enabled we can actually display the BO and use it
- Now the next steps will be to replace and refacto all the parts that are not goo yet

TLDR: most of this is TEMPORARY it will NEVER be released as is, so do you have a better idea to start this long project while still being able to use the BO for a human and apply the UI tests?

File: src/PrestaShopBundle/Bridge/AdminController/LegacyControllerBridge.php [hunk 7]
```diff
@@ -80,6 +80,7 @@
  * @property array $modals
  * @property int $multishop_context
  * @property bool $multishop_context_group
+ * @property string $controller_type
```
Comment (M0rgan01) [inline src/PrestaShopBundle/Bridge/AdminController/LegacyControllerBridge.php:83]:
```suggestion
 * @property string $redirect_after
 * @property bool $lockedToAllShopContext
 * @property string $controller_type
```

File: src/PrestaShopBundle/Bridge/Smarty/LayoutConfigurator.php [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+use Tools;
+
+class LayoutConfigurator implements ConfiguratorInterface
+{
+    public function __construct(
+        private ConfigurationInterface $configuration,
+        private LegacyContext $legacyContext,
+        private TranslatorInterface $translator,
+        private ShopConstraintContextInterface $shopConstraintContext,
+        private TabRepository $tabRepository
```
Comment (M0rgan01) [inline src/PrestaShopBundle/Bridge/Smarty/LayoutConfigurator.php:48]:
```suggestion
        private readonly ConfigurationInterface $configuration,
        private readonly LegacyContext $legacyContext,
        private readonly TranslatorInterface $translator,
        private readonly ShopConstraintContextInterface $shopConstraintContext,
        private readonly TabRepository $tabRepository
```

File: src/PrestaShopBundle/Bridge/SymfonyLayoutFeature.php [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+    public function __construct(
+        FeatureFlagRepository $featureFlagRepository,
+        RequestStack $requestStack,
+        bool $enabledByEnv
+    ) {
+        $this->featureFlagRepository = $featureFlagRepository;
+        $this->requestStack = $requestStack;
+        $this->enabledByEnv = $enabledByEnv;
+    }
```
Comment (M0rgan01) [inline src/PrestaShopBundle/Bridge/SymfonyLayoutFeature.php:64]:
```suggestion
    public function __construct(
        private readonly FeatureFlagRepository $featureFlagRepository,
        private readonly RequestStack $requestStack,
        private readonly bool $enabledByEnv
    ) {
    }
```

File: src/PrestaShopBundle/EventListener/ContextShopListener.php [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+{
+    public const CONTROLLER_CONFIGURATION_ATTRIBUTE = 'configuration_controller';
+
+    public function __construct(
+        private LegacyContext $legacyContext,
+        private Repository $localeRepository,
+        private LegacyControllerBridgeFactory $controllerBridgeFactory,
+        private ControllerConfigurationFactory $controllerConfigurationFactory,
+        private SymfonyLayoutFeature $symfonyLayoutFeature,
+        private ConfigurationInterface $configuration
```
Comment (M0rgan01) [inline src/PrestaShopBundle/EventListener/ContextShopListener.php:54]:
```suggestion
        private readonly LegacyContext $legacyContext,
        private readonly Repository $localeRepository,
        private readonly LegacyControllerBridgeFactory $controllerBridgeFactory,
        private readonly ControllerConfigurationFactory $controllerConfigurationFactory,
        private readonly SymfonyLayoutFeature $symfonyLayoutFeature,
        private readonly ConfigurationInterface $configuration
```

File: src/PrestaShopBundle/Resources/views/Admin/Layout/header.html.twig [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  {% for jsUrl in js_files %}
+    <script type="text/javascript" src="{{ jsUrl }}"></script>
+  {% endfor %}
+{% endif %}
+
+{# Javascript definitions from Media #}
+{% if js_def is not empty %}
+<script type="text/javascript">
+  {% for jsVar, jsValue in js_def %}
+    var {{ jsVar }} = {{ jsValue|json_encode|raw }};
```
Comment (M0rgan01) [inline src/PrestaShopBundle/Resources/views/Admin/Layout/header.html.twig:94]: We can use let, or const if this is not overridden

File: src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+        LegacyContext $context,
+        RequestStack $requestStack,
+        iterable $configurators,
+        SymfonyLayoutFeature $symfonyLayoutFeature
+    ) {
+        $this->context = $context;
+        $this->requestStack = $requestStack;
+        $this->configurators = $configurators;
+        $this->symfonyLayoutFeature = $symfonyLayoutFeature;
+    }
```
Comment (M0rgan01) [inline src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php:82]:
```suggestion
    public function __construct(
        private readonly LegacyContext $context,
        private readonly RequestStack $requestStack,
        private readonly iterable $configurators,
        private readonly SymfonyLayoutFeature $symfonyLayoutFeature
    ) {
    }
```

File: src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+            'use_symfony_layout' => $useSymfonyLayout,
+        ];
+    }
+
+    /**
+     * Returns a list of functions to add to the existing list.
+     *
+     * @return array An array of functions
+     */
+    public function getFunctions()
```
Comment (M0rgan01) [inline src/PrestaShopBundle/Twig/Extension/SymfonyLayoutExtension.php:114]:
```suggestion
    public function getFunctions(): array
```

File: admin-dev/index.php [hunk 14]
```diff
@@ -70,6 +71,17 @@
...
+// Loads .env file from the root of project if present
+$dotEnvFile = dirname(__FILE__, 2) . '/.env';
+if (file_exists($dotEnvFile)) {
+    (new Dotenv())
+        // DO NOT use putEnv
+        ->usePutenv(false)
+        ->loadEnv($dotEnvFile)
+    ;
+}
+
```
Comment (lartist) [inline admin-dev/index.php:84]: What about .env.local, .env.*.local ... ?

File: app/config/config.yml [hunk 15]
```diff
@@ -12,13 +12,16 @@ parameters:
...
+  # This is hard-coded for now but it should be dynamic wince admin folder name changes in prod environments
+  admin_themes_folder: "%kernel.project_dir%/admin-dev/themes"
   admin_page: "%prestashop_views%/Admin"
   env(PS_LOG_OUTPUT): "%kernel.logs_dir%/%kernel.environment%.log"
   env(PS_LOG_MAX_FILES): '30'
   mail_themes_uri: "/mails/themes"
   mail_themes_dir: "%kernel.project_dir%%mail_themes_uri%"
   modules_translation_paths: [ ]
   api_base_path: !php/const PrestaShopBundle\Api\Api::API_BASE_PATH
+  env(USE_SYMFONY_LAYOUT): 'false'
```
Comment (lartist) [inline app/config/config.yml:24]: We should put this in the .env no?

File: src/Adapter/LegacyContext.php [hunk 16]
```diff
@@ -81,7 +90,11 @@ public function getContext()
         if (null === static::$instance) {
             $legacyContext = Context::getContext();
 
-            if ($legacyContext && !empty($legacyContext->shop) && !isset($legacyContext->controller) && isset($legacyContext->employee)) {
+            // This getter initialises a legacy admin controller, in the process of migrating the layout the Context::controller shouldn't be created
+            // by this method, this responsibility lies in the ContextShopListener that create a bridge admin controller, so when Symfony layout is
+            // enabled we skip this step to avoid instantiating multiple legacy controllers
+            $symfonyLayoutEnabled = $this->symfonyLayoutFeature && $this->symfonyLayoutFeature->isEnabled();
```
Comment (lartist) [inline src/Adapter/LegacyContext.php:96]:
```suggestion
            $symfonyLayoutEnabled = $this->symfonyLayoutFeature?->isEnabled();
```

Not blocking :) 

https://php.watch/versions/8.0/null-safe-operator

File: src/PrestaShopBundle/Bridge/AdminController/ControllerConfigurationFactory.php [hunk 17]
```diff
@@ -39,41 +40,24 @@
...
-    /**
-     * @param UserProvider $userProvider
-     * @param Link $link
-     */
     public function __construct(
-        UserProvider $userProvider,
-        Link $link
+        private UserProvider $userProvider,
+        private Link $link,
+        private TabRepository $tabRepository
```
Comment (lartist) [inline src/PrestaShopBundle/Bridge/AdminController/ControllerConfigurationFactory.php:46]:
```suggestion
        private readonly UserProvider $userProvider,
        private readonly Link $link,
        private readonly TabRepository $tabRepository
```

Not blocking

File: src/PrestaShopBundle/Bridge/SymfonyLayoutFeature.php [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+    public function isEnabled(): bool
+    {
+        // When parameter is specified in the url it overrides the other settings, this helps developer debug more easily
+        // by forcing the layout via a query parameter
+        if ($this->requestStack->getCurrentRequest() &&
+            $this->requestStack->getCurrentRequest()->query->has('use_symfony_layout')) {
```
Comment (lartist) [inline src/PrestaShopBundle/Bridge/SymfonyLayoutFeature.php:71]:
```suggestion
        if ($this->requestStack->getCurrentRequest()?->query->has('use_symfony_layout')) {
```
